base/util.go

```diff
@@ -1018,6 +1020,43 @@ func HexCasToUint64(cas string) uint64 {
 	return binary.LittleEndian.Uint64(casBytes[0:8])
 }
 
+// HexCasToUint64ForDelta will convert hex cas to uint64 accounting for any stripped zeros in delta calculation
+func HexCasToUint64ForDelta(casByte []byte) (uint64, error) {
+	if len(casByte) <= 2 {
+		return 0, fmt.Errorf("hex value is too short.")
+	}
+	if casByte[0] != '0' || casByte[1] != 'x' {
+		return 0, fmt.Errorf("incorrect hex value, leading 0x is expected")
+	}
+
+	// as we strip any zeros iff the end of the hex value for deltas, the input delta could be odd length
+	if len(casByte)%2 != 0 {
+		evenHexLen := make([]byte, len(casByte), len(casByte)+1)
+		copy(evenHexLen, casByte)
+		evenHexLen = append(evenHexLen, '0')
+		casByte = evenHexLen
+	}
+
+	decoded := make([]byte, MaxDecodedLength)
+	_, err := hex.Decode(decoded, casByte[2:])
+	if err != nil {
+		return 0, err
+	}
+	res := binary.LittleEndian.Uint64(decoded)
```

Comment:
Doesn't binary.LittleEndian.Uint64 expect a length 8 []byte as input?  Here it looks like we're setting decoded to MaxDecodedLength (18) - can you help me understand the case where casByte[2:] is longer than 8 bytes but we need to support this and are intentionally ignoring trailing bytes?

I thought this was just handling the case where it's less than 8 bytes because we've trimmed trailing zeros.

Reply:
Good point. I don't think the max decoded length line is needed, I have removed it.

Reply:
Well, hex.Decode is going to panic if casByte[2:] overflows decoded, so it's possible the previous code was defensive.  https://go.dev/play/p/_LEScaOT1FG
Our HexCasToUint64 has an explicit length check to protect against this.  Possibly we just want to do the same (but also supporting the case here that the length can be less than 8): doing something like return error `if len(casByte) > 10` at the top of the function?

Reply:
Added some safety checks around this area, let me know what you think